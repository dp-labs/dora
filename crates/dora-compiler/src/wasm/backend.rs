use dora_ir::IRTypes;
use dora_runtime::symbols;
use melior::dialect::{
    arith::{self, CmpfPredicate, CmpiPredicate},
    cf, func,
};
use melior::ir::{
    attribute::FlatSymbolRefAttribute, r#type::IntegerType, BlockRef, OperationRef, Region, Type,
    TypeLike, Value, ValueLike,
};
use wasmer_types::TrapCode;

use crate::backend::IntCC;
use crate::errors::Result;
use crate::value::ToContextValue;
use crate::{backend::TypeMethods, context::Context, conversion::builder::OpBuilder, state::State};

use super::intrinsics::WASMIntrinsics;

/// A builder structure for constructing WebAssembly (WASM) operations and managing
/// the underlying compilation process using an `OpBuilder`. The `WASMBuilder`
/// operates on a `WASMBackend` and provides an interface to emit operations
/// within the context of a given WASM execution model.
///
/// # Fields:
/// - `backend`: A mutable reference to the `WASMBackend`, which contains the overall
///   context and state of the WASM environment being built.
/// - `builder`: An `OpBuilder` instance responsible for generating operations
///   during the construction process.
///
/// # Example Usage:
/// ```no_check
/// let mut wasm_builder = WASMBuilder {
///     backend: &mut wasm_backend,
///     builder: op_builder,
/// };
/// wasm_builder.builder.some_build_method();
/// ```
///
/// # Notes:
/// - The `WASMBuilder` works closely with the `WASMBackend` to manage WebAssembly-specific
///   context and state, allowing for the construction of WASM operations within the MLIR system.
#[derive(Debug)]
pub struct WASMBuilder<'c, 'a> {
    /// A mutable reference to the WASM backend containing the context and state.
    pub backend: &'a mut WASMBackend<'c>,
    /// The builder used to generate operations within the WASM execution context.
    pub builder: OpBuilder<'c, 'a>,
}

impl<'c> std::ops::Deref for WASMBuilder<'c, '_> {
    type Target = WASMBackend<'c>;

    #[inline]
    fn deref(&self) -> &Self::Target {
        self.backend
    }
}

impl std::ops::DerefMut for WASMBuilder<'_, '_> {
    #[inline]
    fn deref_mut(&mut self) -> &mut Self::Target {
        self.backend
    }
}

/// The backend structure for managing the WebAssembly (WASM) execution context and
/// intrinsics. `WASMBackend` encapsulates the relevant MLIR context, intrinsic types,
/// and execution state required for compiling and executing WASM bytecode.
///
/// # Fields:
/// - `ctx`: A reference to the MLIR `Context`, which provides the underlying infrastructure
///   for managing the compilation and execution process.
/// - `intrinsics`: A set of WebAssembly-specific intrinsic types, such as i32, i64, f32, and f64,
///   that are commonly used in WASM operations.
/// - `state`: The current state of the execution, which includes the stack and control
///   structures required for managing the flow of WASM bytecode execution.
///
/// # Example Usage:
/// ```no_check
/// let wasm_backend = WASMBackend {
///     ctx: &mlir_context,
///     intrinsics: wasm_intrinsics,
///     state: wasm_state,
/// };
/// ```
///
/// # Notes:
/// - The `WASMBackend` is responsible for maintaining the state and context necessary
///   for WASM execution and compilation. It interacts with other structures like the
///   `WASMBuilder` to facilitate the construction of operations that adhere to the
///   WebAssembly execution model.
#[derive(Debug)]
pub struct WASMBackend<'c> {
    /// A reference to the MLIR context used in the WASM backend.
    pub ctx: &'c Context,
    /// WebAssembly-specific intrinsic types, such as i32, i64, f32, and f64.
    pub intrinsics: WASMIntrinsics<'c>,
    /// The current state of WASM execution, including the stack and control structures.
    pub state: State<'c, 'c>,
}

impl<'c> WASMBackend<'c> {
    pub fn new(ctx: &'c Context) -> Self {
        Self {
            ctx,
            intrinsics: WASMIntrinsics::declare(ctx),
            state: State::new(),
        }
    }
}

impl<'c> IRTypes for WASMBackend<'c> {
    type Type = Type<'c>;
    type Value = Value<'c, 'c>;
    type Region = Region<'c>;
    type BasicBlock = BlockRef<'c, 'c>;
    type Operation = OperationRef<'c, 'c>;
}

impl TypeMethods for WASMBackend<'_> {
    fn type_ptr(&self) -> Self::Type {
        self.intrinsics.ptr_ty
    }

    fn type_int(&self, bits: u32) -> Self::Type {
        IntegerType::new(&self.ctx.mlir_context, bits).into()
    }
}

pub(crate) fn is_zero<'c, 'a>(
    builder: &OpBuilder<'c, 'a>,
    value: Value<'c, 'a>,
) -> Result<Value<'c, 'a>> {
    let ty = value.r#type();
    if ty.is_integer() {
        Ok(builder
            .create(arith::cmpi(
                builder.context(),
                CmpiPredicate::Eq,
                value,
                if builder.int_ty_width(ty)? == 32 {
                    builder.create(builder.iconst_32(0)).result(0)?.into()
                } else {
                    builder.create(builder.iconst_64(0)).result(0)?.into()
                },
                builder.unknown_loc(),
            ))
            .result(0)?
            .to_ctx_value())
    } else if ty.is_float() {
        Ok(builder
            .create(arith::cmpf(
                builder.context(),
                CmpfPredicate::Oeq,
                value,
                if ty.is_f32() {
                    builder.create(builder.fconst_32(0.0)).result(0)?.into()
                } else {
                    builder.create(builder.fconst_64(0.0)).result(0)?.into()
                },
                builder.unknown_loc(),
            ))
            .result(0)?
            .to_ctx_value())
    } else {
        // WASM pointer type
        let value = builder.make(builder.load(value, builder.i32_ty()))?;
        let result = builder.make(builder.icmp_imm(IntCC::Equal, value, 0)?)?;
        Ok(unsafe { Value::from_raw(result.to_raw()) })
    }
}

pub(crate) fn trap<'c, 'a>(
    builder: &OpBuilder<'c, 'a>,
    code: TrapCode,
    continue_block: BlockRef<'c, 'a>,
) -> Result<()> {
    let ctx = builder.ctx;
    let code = builder.make(builder.iconst_32(code as _))?;
    builder.create(func::call(
        ctx,
        FlatSymbolRefAttribute::new(ctx, symbols::wasm::RAISE_TRAP),
        &[code],
        &[],
        builder.get_insert_location(),
    ));
    builder.create(cf::br(&continue_block, &[], builder.get_insert_location()));
    Ok(())
}
