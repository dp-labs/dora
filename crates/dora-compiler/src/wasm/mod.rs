//! Reference: https://github.com/wasmerio/wasmer/tree/main/lib/compiler-llvm
#![allow(dead_code)]

mod backend;
mod code;
mod errors;
mod func;
mod intrinsics;
mod state;
mod ty;

#[cfg(test)]
mod tests;

use dora_primitives::config::OptimizationLevel;
use func::FuncTranslator;
use melior::ir::operation::OperationBuilder;
use melior::ir::{Block, Region};
use melior::ir::{Location, Module as MLIRModule};
use std::sync::Arc;
use wasmer_compiler::{
    FunctionBodyData, ModuleEnvironment, ModuleMiddleware, ModuleTranslationState,
};
use wasmer_types::entity::{EntityRef, PrimaryMap};
use wasmer_types::{
    CompileModuleInfo, Features, FunctionIndex, LocalFunctionIndex, SectionIndex, SignatureIndex,
    Symbol, SymbolRegistry,
};

use crate::errors::CompileError;
use crate::{Context, Module};

/// Represents a WebAssembly (Wasm) to LLVM/MLIR compiler. The `WASMCompiler` struct is responsible for
/// compiling WebAssembly code into LLVM/MLIR, leveraging the configuration provided at the time of
/// instantiation.
///
/// # Fields:
/// - `config`: The configuration settings for the compiler, encapsulated in the `Config` struct.
///
/// # Example Usage:
/// ```no_check
/// let config = Config {
///     enable_nan_canonicalization: true,
///     enable_verifier: false,
///     opt_level: OptimizationLevel::O2,
///     middlewares: vec![],
/// };
/// let wasm_compiler = WASMCompiler::new(config);
///
/// // Access the compiler's configuration
/// let compiler_config = wasm_compiler.config();
/// ```
///
/// # Notes:
/// - The `WASMCompiler` struct encapsulates all compiler functionality for transforming WebAssembly bytecode
///   into a more optimized and executable format using LLVM/MLIR.
/// - The compiler that compiles a WebAssembly module with LLVM/MLIR, translating the Wasm to MLIR IR, optimizing
///   it and then translating to assembly or executing it with JIT mode.
#[derive(Debug, Clone)]
pub struct WASMCompiler<'c> {
    /// The MLIR context used for generating operations and managing their lifetime. It encapsulates the state
    /// of the MLIR infrastructure, including types, modules, and operations. This context is tied to the
    /// lifetime `'c` of the EVMCompiler.
    pub ctx: &'c Context,
    /// The configuration settings for the Wasm compiler.
    pub config: Config,
}

impl<'c> WASMCompiler<'c> {
    /// Creates a new instance of the `WASMCompiler` with the given configuration.
    ///
    /// # Arguments:
    /// - `config`: The `Config` object that defines the compilation settings, including optimization levels
    ///   and any middleware.
    ///
    /// # Returns:
    /// - A `WASMCompiler` instance initialized with the provided configuration.
    #[inline]
    pub fn new(ctx: &'c Context, config: Config) -> WASMCompiler {
        WASMCompiler { ctx, config }
    }

    /// Retrieves the configuration for this `WASMCompiler`.
    ///
    /// # Returns:
    /// - A reference to the `Config` object that was used to configure the compiler.
    ///
    /// # Example:
    /// ```no_check
    /// let compiler_config = wasm_compiler.config();
    /// ```
    #[inline]
    fn config(&self) -> &Config {
        &self.config
    }

    /// Compile the WASM bytes using LLVM/MLIR.
    pub fn compile(&self, data: &[u8]) -> Result<Module<'c>, CompileError> {
        let environ = ModuleEnvironment::new();
        let translation = environ
            .translate(data)
            .map_err(|err| CompileError::Codegen(err.to_string()))?;
        let module = translation.module;
        let compile_info = CompileModuleInfo {
            module: Arc::new(module),
            features: Features::default(),
            memory_styles: PrimaryMap::new(),
            table_styles: PrimaryMap::new(),
        };
        self.compile_module(
            &compile_info,
            translation.module_translation_state.as_ref().unwrap(),
            translation.function_body_inputs,
        )
    }

    fn compile_module(
        &self,
        compile_info: &CompileModuleInfo,
        module_translation: &ModuleTranslationState,
        function_body_inputs: PrimaryMap<LocalFunctionIndex, FunctionBodyData<'_>>,
    ) -> Result<Module<'c>, CompileError> {
        // WASM Module information
        let memory_styles = &compile_info.memory_styles;
        let table_styles = &compile_info.table_styles;
        let wasm_module = &compile_info.module;
        let function_body_inputs = function_body_inputs
            .iter()
            .collect::<Vec<(LocalFunctionIndex, &FunctionBodyData<'_>)>>();
        let functions: Vec<_> = function_body_inputs
            .iter()
            .map(|(i, input)| {
                FuncTranslator::translate(
                    self.ctx,
                    &self.config,
                    wasm_module,
                    module_translation,
                    i,
                    input,
                    memory_styles,
                    table_styles,
                    &ShortNames {},
                )
            })
            .collect();
        // MLIR Module region and block
        let module_block = Block::new(&[]);
        for func in functions {
            module_block
                .append_operation(func.map_err(|err| CompileError::Codegen(err.to_string()))?);
        }
        let module_region = Region::new();
        module_region.append_block(module_block);
        // Build main module
        let op = OperationBuilder::new("builtin.module", Location::unknown(&self.ctx.mlir_context))
            .add_regions([module_region])
            .build()?;
        Ok(Module::new(
            MLIRModule::from_operation(op).expect("module failed to create"),
        ))
    }
}

struct ShortNames {}

impl SymbolRegistry for ShortNames {
    fn symbol_to_name(&self, symbol: Symbol) -> String {
        match symbol {
            Symbol::Metadata => "M".to_string(),
            Symbol::LocalFunction(index) => format!("f{}", index.index()),
            Symbol::Section(index) => format!("s{}", index.index()),
            Symbol::FunctionCallTrampoline(index) => format!("t{}", index.index()),
            Symbol::DynamicFunctionTrampoline(index) => format!("d{}", index.index()),
        }
    }

    fn name_to_symbol(&self, name: &str) -> Option<Symbol> {
        if name.len() < 2 {
            return None;
        }
        let (ty, idx) = name.split_at(1);
        if ty.starts_with('M') {
            return Some(Symbol::Metadata);
        }

        let idx = idx.parse::<u32>().ok()?;
        match ty.chars().next().unwrap() {
            'f' => Some(Symbol::LocalFunction(LocalFunctionIndex::from_u32(idx))),
            's' => Some(Symbol::Section(SectionIndex::from_u32(idx))),
            't' => Some(Symbol::FunctionCallTrampoline(SignatureIndex::from_u32(
                idx,
            ))),
            'd' => Some(Symbol::DynamicFunctionTrampoline(FunctionIndex::from_u32(
                idx,
            ))),
            _ => None,
        }
    }
}

/// Represents the configuration options for the `WASMCompiler`. The `Config` struct defines various
/// settings that influence the compilation process, such as optimization levels and verification options.
///
/// # Fields:
/// - `enable_nan_canonicalization`: A boolean flag indicating whether to enable NaN canonicalization, which
///   standardizes NaN representations in floating-point operations.
/// - `enable_verifier`: A boolean flag to enable or disable the verification phase during compilation, ensuring
///   that the compiled code meets correctness constraints.
/// - `opt_level`: The optimization level to be applied during compilation, which can affect the performance and
///   size of the compiled output.
/// - `middlewares`: A vector of middleware components that modify or extend the behavior of the compilation
///   process. Each middleware is represented as a trait object implementing the `ModuleMiddleware` trait.
///
/// # Example Usage:
/// ```no_check
/// let config = Config {
///     enable_nan_canonicalization: true,
///     enable_verifier: true,
///     opt_level: OptimizationLevel::O3,
///     middlewares: vec![],
/// };
/// ```
///
/// # Notes:
/// - The `Config` struct is essential for controlling the behavior of the `WASMCompiler`. You can tune the
///   compilation process by adjusting these fields based on the requirements of your WebAssembly module.
#[derive(Debug, Clone, Default)]
pub struct Config {
    /// A flag to enable NaN canonicalization during floating-point operations.
    pub(crate) enable_nan_canonicalization: bool,
    /// A flag to enable or disable the verifier during compilation.
    pub(crate) enable_verifier: bool,
    /// The optimization level for the compilation process.
    pub(crate) opt_level: OptimizationLevel,
    /// A collection of middleware components that modify the behavior of the compiler.
    pub(crate) middlewares: Vec<Arc<dyn ModuleMiddleware>>,
}
