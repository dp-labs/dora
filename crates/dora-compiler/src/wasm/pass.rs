use super::conversion;
use crate::errors::Result;
use conversion::ConversionPass;
use melior::{ir::Module as MLIRModule, Context};

/// Define a public function `run` to execute the conversion process.
///
/// # Parameters
/// - `ctx`: An MLIR context, used to manage the global state of MLIR.
/// - `module`: A mutable MLIR module representing the MLIR code to be transformed.
///
/// # Returns
/// - `Result<()>`, indicating success or failure of the operation.
pub fn run(ctx: &Context, module: &mut MLIRModule) -> Result<()> {
    let mut conversion_pass = ConversionPass { ctx };
    conversion_pass.run(module.as_operation())
}
