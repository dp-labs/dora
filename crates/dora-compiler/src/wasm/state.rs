use crate::state::ExtraInfo;
use std::ops::{BitAnd, BitOr, BitOrAssign};

impl ExtraInfo {
    // This value is required to be arithmetic 32-bit NaN (or 32x4) by the WAsm
    // machine, but which might not be in the LLVM value. The conversion to
    // arithmetic NaN is pending. It is required for correctness.
    //
    // When applied to a 64-bit value, this flag has no meaning and must be
    // ignored. It may be set in such cases to allow for common handling of
    // 32 and 64-bit operations.
    pub const fn pending_f32_nan() -> ExtraInfo {
        ExtraInfo { state: 1 }
    }

    // This value is required to be arithmetic 64-bit NaN (or 64x2) by the WAsm
    // machine, but which might not be in the LLVM value. The conversion to
    // arithmetic NaN is pending. It is required for correctness.
    //
    // When applied to a 32-bit value, this flag has no meaning and must be
    // ignored. It may be set in such cases to allow for common handling of
    // 32 and 64-bit operations.
    pub const fn pending_f64_nan() -> ExtraInfo {
        ExtraInfo { state: 2 }
    }

    // This value either does not contain a 32-bit NaN, or it contains an
    // arithmetic NaN. In SIMD, applies to all 4 lanes.
    pub const fn arithmetic_f32() -> ExtraInfo {
        ExtraInfo { state: 4 }
    }

    // This value either does not contain a 64-bit NaN, or it contains an
    // arithmetic NaN. In SIMD, applies to both lanes.
    pub const fn arithmetic_f64() -> ExtraInfo {
        ExtraInfo { state: 8 }
    }

    pub const fn has_pending_f32_nan(&self) -> bool {
        self.state & ExtraInfo::pending_f32_nan().state != 0
    }
    pub const fn has_pending_f64_nan(&self) -> bool {
        self.state & ExtraInfo::pending_f64_nan().state != 0
    }
    pub const fn is_arithmetic_f32(&self) -> bool {
        self.state & ExtraInfo::arithmetic_f32().state != 0
    }
    pub const fn is_arithmetic_f64(&self) -> bool {
        self.state & ExtraInfo::arithmetic_f64().state != 0
    }

    pub const fn strip_pending(&self) -> ExtraInfo {
        ExtraInfo {
            state: self.state
                & !(ExtraInfo::pending_f32_nan().state | ExtraInfo::pending_f64_nan().state),
        }
    }
}

// Union two ExtraInfos.
impl BitOr for ExtraInfo {
    type Output = Self;

    fn bitor(self, other: Self) -> Self {
        debug_assert!(!(self.has_pending_f32_nan() && other.has_pending_f64_nan()));
        debug_assert!(!(self.has_pending_f64_nan() && other.has_pending_f32_nan()));
        ExtraInfo {
            state: if self.is_arithmetic_f32() || other.is_arithmetic_f32() {
                ExtraInfo::arithmetic_f32().state
            } else if self.has_pending_f32_nan() || other.has_pending_f32_nan() {
                ExtraInfo::pending_f32_nan().state
            } else {
                0
            } + if self.is_arithmetic_f64() || other.is_arithmetic_f64() {
                ExtraInfo::arithmetic_f64().state
            } else if self.has_pending_f64_nan() || other.has_pending_f64_nan() {
                ExtraInfo::pending_f64_nan().state
            } else {
                0
            },
        }
    }
}
impl BitOrAssign for ExtraInfo {
    fn bitor_assign(&mut self, other: Self) {
        *self = *self | other;
    }
}

// Intersection for ExtraInfo.
impl BitAnd for ExtraInfo {
    type Output = Self;
    fn bitand(self, other: Self) -> Self {
        debug_assert!(
            self.has_pending_f32_nan() == other.has_pending_f32_nan()
                || self.is_arithmetic_f32()
                || other.is_arithmetic_f32()
        );
        debug_assert!(
            self.has_pending_f64_nan() == other.has_pending_f64_nan()
                || self.is_arithmetic_f64()
                || other.is_arithmetic_f64()
        );
        let info = match (
            self.is_arithmetic_f32() && other.is_arithmetic_f32(),
            self.is_arithmetic_f64() && other.is_arithmetic_f64(),
        ) {
            (false, false) => Default::default(),
            (true, false) => ExtraInfo::arithmetic_f32(),
            (false, true) => ExtraInfo::arithmetic_f64(),
            (true, true) => ExtraInfo::arithmetic_f32() | ExtraInfo::arithmetic_f64(),
        };
        match (self.has_pending_f32_nan(), self.has_pending_f64_nan()) {
            (false, false) => info,
            (true, false) => info | ExtraInfo::pending_f32_nan(),
            (false, true) => info | ExtraInfo::pending_f64_nan(),
            (true, true) => unreachable!("Can't form ExtraInfo"),
        }
    }
}
