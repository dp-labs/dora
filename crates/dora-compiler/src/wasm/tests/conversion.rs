//！Reference: https://github.com/WebAssembly/spec/tree/main/test/core
use crate::context::Context;
use crate::wasm::{Config, WASMCompiler};
use wasmer::wat2wasm;

macro_rules! assert_snapshot {
    ($code:expr) => {
        let context = Context::new();
        let compiler = WASMCompiler::new(&context, Config::default());
        let wasm_bytes = wat2wasm($code.as_bytes()).unwrap();
        // Compile WASM Bytecode to MLIR EVM Dialect
        let mut module = compiler
            .compile(&wasm_bytes)
            .expect("failed to compile program");
        // Lowering the WASM dialect to the Dora dialect.
        crate::wasm::pass::run(&context.mlir_context, &mut module.mlir_module)
            .expect("failed to run conversion pass");
        // Lowering the Dora dialect to MLIR builtin dialects.
        crate::dora::pass::run(
            &context.mlir_context,
            &mut module.mlir_module,
            &crate::dora::pass::PassOptions {
                code_size: wasm_bytes.len() as u32,
                ..Default::default()
            },
        )
        .unwrap();
        let op = module.module().as_operation();
        assert!(op.verify());
        insta::assert_snapshot!(op);
    };
}

#[test]
fn unreachable() {
    assert_snapshot!(
        r#"
(module
    (func (export "break-bare") (result i32)
    (block (br 0) (unreachable))
    (block (br_if 0 (i32.const 1)) (unreachable))
    (block (br_table 0 (i32.const 0)) (unreachable))
    (block (br_table 0 0 0 (i32.const 1)) (unreachable))
    (i32.const 19)
  )
)
"#
    );
}

#[test]
fn sum() {
    assert_snapshot!(
        r#"
(module
  (type $sum_t (func (param i32 i32) (result i32)))
  (func $sum_f (type $sum_t) (param $x i32) (param $y i32) (result i32)
    local.get $x
    local.get $y
    i32.add)
  (export "sum" (func $sum_f)))
"#
    );
}

#[test]
fn sum_with_export_functions() {
    assert_snapshot!(
        r#"
(module
  (func $sum_f (param $x i32) (param $y i32) (result i32)
    local.get $x
    local.get $y
    i32.add)

  (func (export "main") 
    (call $sum_f (i32.const 2) (i32.const 3))
  )
)
"#
    );
}

#[test]
fn and_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "and") (param $x i32) (param $y i32) (result i32) (i32.and (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn and_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "and") (param $x i64) (param $y i64) (result i64) (i64.and (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn or_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "or") (param $x i32) (param $y i32) (result i32) (i32.or (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn or_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "or") (param $x i64) (param $y i64) (result i64) (i64.or (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn xor_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "xor") (param $x i32) (param $y i32) (result i32) (i32.xor (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn xor_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "xor") (param $x i64) (param $y i64) (result i64) (i64.xor (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shl_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "shl") (param $x i32) (param $y i32) (result i32) (i32.shl (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shl_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "shl") (param $x i64) (param $y i64) (result i64) (i64.shl (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shr_s_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "shr_s") (param $x i32) (param $y i32) (result i32) (i32.shr_s (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shr_s_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "shr_s") (param $x i64) (param $y i64) (result i64) (i64.shr_s (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shr_u_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "shr_u") (param $x i32) (param $y i32) (result i32) (i32.shr_u (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn shr_u_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "shr_u") (param $x i64) (param $y i64) (result i64) (i64.shr_u (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn rotl_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "rotl") (param $x i32) (param $y i32) (result i32) (i32.rotl (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn rotl_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "rotl") (param $x i64) (param $y i64) (result i64) (i64.rotl (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn rotr_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "rotr") (param $x i32) (param $y i32) (result i32) (i32.rotr (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn rotr_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "rotr") (param $x i64) (param $y i64) (result i64) (i64.rotr (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn clz_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "clz") (param $x i32) (result i32) (i32.clz (local.get $x)))
)
"#
    );
}

#[test]
fn clz_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "clz") (param $x i64) (result i64) (i64.clz (local.get $x)))
)
"#
    );
}

#[test]
fn ctz_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "ctz") (param $x i32) (result i32) (i32.ctz (local.get $x)))
)
"#
    );
}

#[test]
fn ctz_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "ctz") (param $x i64) (result i64) (i64.ctz (local.get $x)))
)
"#
    );
}

#[test]
fn popcnt_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "popcnt") (param $x i32) (result i32) (i32.popcnt (local.get $x)))
)
"#
    );
}

#[test]
fn popcnt_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "popcnt") (param $x i64) (result i64) (i64.popcnt (local.get $x)))
)
"#
    );
}

#[test]
fn extend_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "extend8_s") (param $x i32) (result i32) (i32.extend8_s (local.get $x)))
  (func (export "extend16_s") (param $x i32) (result i32) (i32.extend16_s (local.get $x)))
)
"#
    );
}

#[test]
fn extend_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "extend8_s") (param $x i64) (result i64) (i64.extend8_s (local.get $x)))
  (func (export "extend16_s") (param $x i64) (result i64) (i64.extend16_s (local.get $x)))
  (func (export "extend32_s") (param $x i64) (result i64) (i64.extend32_s (local.get $x)))
)
"#
    );
}

#[test]
fn eq_i32() {
    assert_snapshot!(
        r#"
(module
  (func (export "eqz") (param $x i32) (result i32) (i32.eqz (local.get $x)))
  (func (export "eq") (param $x i32) (param $y i32) (result i32) (i32.eq (local.get $x) (local.get $y)))
  (func (export "ne") (param $x i32) (param $y i32) (result i32) (i32.ne (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn eq_i64() {
    assert_snapshot!(
        r#"
(module
  (func (export "eqz") (param $x i64) (result i32) (i64.eqz (local.get $x)))
  (func (export "eq") (param $x i64) (param $y i64) (result i32) (i64.eq (local.get $x) (local.get $y)))
  (func (export "ne") (param $x i64) (param $y i64) (result i32) (i64.ne (local.get $x) (local.get $y)))
)
"#
    );
}

#[test]
fn select() {
    assert_snapshot!(
        r#"
(module
    (func $select (param $cond i32) (param $a i32) (param $b i32) (result i32)
        local.get $cond
        local.get $a
        local.get $b
        select
    )
    (export "select" (func $select))
)
"#
    );
}
