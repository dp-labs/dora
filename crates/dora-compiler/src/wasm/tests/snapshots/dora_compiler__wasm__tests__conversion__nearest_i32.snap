---
source: crates/dora-compiler/src/wasm/tests/conversion.rs
expression: op
snapshot_kind: text
---
module {
  func.func public @nearest(%arg0: !llvm.ptr, %arg1: f32) -> f32 attributes {llvm.emit_c_interface} {
    %c1_i32 = arith.constant 1 : i32
    %0 = llvm.alloca %c1_i32 x f32 : (i32) -> !llvm.ptr
    llvm.store %arg1, %0 : f32, !llvm.ptr
    cf.br ^bb1
  ^bb1:  // pred: ^bb0
    %1 = llvm.load %0 : !llvm.ptr -> f32
    %2 = arith.negf %1 : f32
    cf.br ^bb2(%2 : f32)
  ^bb2(%3: f32):  // pred: ^bb1
    return %3 : f32
  }
  func.func private @dora_fn_wasm_table_init(!llvm.ptr, i32, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_table_copy(!llvm.ptr, i32, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_table_fill(!llvm.ptr, i32, i32, !llvm.ptr, i32)
  func.func private @dora_fn_wasm_table_size(!llvm.ptr, i32) -> i32
  func.func private @dora_fn_wasm_table_get(!llvm.ptr, i32, i32) -> !llvm.ptr
  func.func private @dora_fn_wasm_table_set(!llvm.ptr, i32, i32, !llvm.ptr)
  func.func private @dora_fn_wasm_table_grow(!llvm.ptr, !llvm.ptr, i32, i32) -> i32
  func.func private @dora_fn_wasm_imported_table_size(!llvm.ptr, i32) -> i32
  func.func private @dora_fn_wasm_imported_table_get(!llvm.ptr, i32, i32) -> !llvm.ptr
  func.func private @dora_fn_wasm_imported_table_grow(!llvm.ptr, !llvm.ptr, i32, i32) -> i32
  func.func private @dora_fn_wasm_memory_init(!llvm.ptr, i32, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_memory_size(!llvm.ptr, i32) -> i32
  func.func private @dora_fn_wasm_memory_grow(!llvm.ptr, i32, i32) -> i32
  func.func private @dora_fn_wasm_memory_copy(!llvm.ptr, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_memory_fill(!llvm.ptr, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_memory_notify(!llvm.ptr, i32, i32, i32) -> i32
  func.func private @dora_fn_wasm_memory_wait32(!llvm.ptr, i32, i32, i32, i64) -> i32
  func.func private @dora_fn_wasm_memory_wait64(!llvm.ptr, i32, i32, i64, i64) -> i32
  func.func private @dora_fn_wasm_imported_memory_size(!llvm.ptr, i32) -> i32
  func.func private @dora_fn_wasm_imported_memory_grow(!llvm.ptr, i32, i32) -> i32
  func.func private @dora_fn_wasm_imported_memory_copy(!llvm.ptr, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_imported_memory_fill(!llvm.ptr, i32, i32, i32, i32)
  func.func private @dora_fn_wasm_imported_memory_notify(!llvm.ptr, i32, i32, i32) -> i32
  func.func private @dora_fn_wasm_imported_memory_wait32(!llvm.ptr, i32, i32, i32, i64) -> i32
  func.func private @dora_fn_wasm_imported_memory_wait64(!llvm.ptr, i32, i32, i64, i64) -> i32
  func.func private @dora_fn_wasm_func_ref(!llvm.ptr, i32) -> !llvm.ptr
  func.func private @dora_fn_wasm_data_drop(!llvm.ptr, i32)
  func.func private @dora_fn_wasm_elem_drop(!llvm.ptr, i32)
  func.func private @dora_fn_wasm_raise_trap(i32)
}
